### ISSUE TITLE
more more firehose issues: rotErr Error

### ISSUE THREAD

**Issue by osrf-migration** [osrf/drcsim#282]
**Original report ([archived issue](https://osrf-migration.github.io/drcsim-gh-pages/#!/osrf/drcsim/issues/282)) by Chris Atkeson (Bitbucket: [catkeson](https://bitbucket.org/%7B3bdaa162-b747-4fa9-9d6e-a2d5c92933c6%7D/)).**

----------------------------------------
I believe the rotation error calculation in the firehose alignment does not currently make sense. I printed out a bunch of stuff at the beginning of VRC_TASK_3:

* Dbg coupling -2.20013 -4.69832 1.06076 -0.003349 0.069256 1.57054
* Dbg spout -2.5 -3.5 1.2 -1.5707 -0 -0
* Dbg connectPose 1.2e-05 -0.125623 0.35 -0.041215 -1.57078 1.61199
* Dbg relativePose 0.299871 0.139127 -1.19833 0.000399 -1.50144 1.56706
* Dbg connect offset -1.54832 -0.299852 -0.264782 -0.003327 0.069336 -0.000259
* Dbg xyz 1.59917
* Dbg rpy 0.0694017
* Dbg valve -1.21592e-12
* Dbg relPose.Zaxis -0.003331 -0.99759 0.069297
* Dbg connPose.Zaxis -2.2e-05 -1 1.6e-05
* Dbg coupling.Zaxis -0.003331 0.069201 0.997597
* Dbg spout.Zaxis 0 1 9.6e-05
* Dbg rotErr_cga 1.36433

At the start the hose points away from the spigot, so having a small rotErr
is bad (0.0694017). The rotErr should be maximal.

The main issue is that the "Z axis" of the hose points upward along the Z axis.
It should point horizontally along the long axis of the coupling, as the hose is horizontal at the start of the task. I see that coupling_relative_pose in the world file
compensates for this. I don't think it is correct. It means the robot needs to very carefully never rotate the coupling about its long axis during the connection. This conflicts with screwing the coupling by rotating it around its long axis.
It also means that any orientation of the coupling with that "Z axis" aligned with
the z axis is okay (including pointing away from the spout, or being perpendicular to it.

I believe this is what is preventing attachment when it appears the hose is correctly aligned.

I think the hose coupling should be redefined so it's Z axis is along its long axis.

Here are some printouts with the coupling in what we believe is roughly the right place and alignment.
Note that coupling.Zaxis is very different in the two runs.

* coupling -3.18106 -4.10214 1.36715 0.621199 0.009171 -1.56043
* spout -3.18 -4.09 1.37 -1.5707 0 -0
* connectPose 1.2e-05 0 0 2.97 1.48 -0.275
* relativePose -0.001061 0.002849 -0.01214 2.91118 1.55702 0.719231
* connect offset 0.011926 -0.00291 0.002439 -1.0515 -0.080077 -0.025667
* xyz 0.0125155
* rpy 1.00531
* valve -1.2168e-12
* relPose.Zaxis -0.581902 -0.813148 -0.013411
* connPose.Zaxis -0.990752 0.102121 -0.08934
* coupling.Zaxis -0.581902 -0.013489 0.813147
* spout.Zaxis 0 1 9.6e-05
* rotErr_cga 1.42366

and

* coupling -3.1951 -4.10624 1.37255 1.74921 0.016014 -1.37723
* spout -3.18 -4.09 1.37 -1.5707 -0 -0
* connectPose 1.2e-05 0 0 2.97 1.48 1.075
* relativePose -0.015105 -0.002557 -0.016241 -2.88217 1.37658 0.082577
* connect offset 0.015318 -0.01376 0.008652 1.41509 -0.01173 0.162707
* xyz 0.0223347
* rpy 1.29995
* valve -1.2168e-12
* relPose.Zaxis -0.966294 0.177429 -0.186535
* connPose.Zaxis -0.316623 -0.944335 -0.08934
* coupling.Zaxis -0.966294 -0.186518 -0.177447
* spout.Zaxis 0 1 9.6e-05
* rotErr_cga 1.54048

Here is the relevant code. I thought rotErr_cga would be the right answer, but it is not given the definition of the axes. Probably the X or Y axis for the coupling link would have worked.

```
#!c++
  double rotErr_cga
    = (this->drcFireHose.couplingLink->GetWorldPose().rot.GetZAxis() -
       this->drcFireHose.spoutLink->GetWorldPose().rot.GetZAxis()).GetLength();

      gzdbg << "\ncoupling " << this->drcFireHose.couplingLink->GetWorldPose() \
<< "\n";
      gzdbg << "spout " << this->drcFireHose.spoutLink->GetWorldPose() << "\n";
      gzdbg << "connectPose " << connectPose << "\n";
      gzdbg << "relativePose " << relativePose << "\n";
      gzdbg << "connect offset " << connectOffset << "\n";
      gzdbg << "xyz " << posErr << "\n";
      gzdbg << "rpy " << rotErr << "\n";
      gzdbg << "valve " << valveAng << "\n";
      gzdbg << "relPose.Zaxis " << relativePose.rot.GetZAxis() << "\n";
      gzdbg << "connPose.Zaxis " << connectPose.rot.GetZAxis() << "\n";
      gzdbg << "coupling.Zaxis " << this->drcFireHose.couplingLink->GetWorldPos\
e().rot.GetZAxis() << "\n";
      gzdbg << "spout.Zaxis " << this->drcFireHose.spoutLink->GetWorldPose().ro\
t.GetZAxis() << "\n";
      gzdbg << "rotErr_cga " << rotErr_cga << "\n";


```

**Comment by osrf-migration** [comment]
**Original comment by Chris Atkeson (Bitbucket: [catkeson](https://bitbucket.org/%7B3bdaa162-b747-4fa9-9d6e-a2d5c92933c6%7D/)).**

----------------------------------------

* Edited issue description

**Comment by osrf-migration** [comment]
**Original comment by Chris Atkeson (Bitbucket: [catkeson](https://bitbucket.org/%7B3bdaa162-b747-4fa9-9d6e-a2d5c92933c6%7D/)).**

----------------------------------------

* Edited issue description

**Comment by osrf-migration** [comment]
**Original comment by david_kortenkamp (Bitbucket: [david_kortenkamp](https://bitbucket.org/%7Bc3ae0326-026c-4c6d-baca-63ac2ebb3f54%7D/)).**

----------------------------------------
We've gotten the hose close to the spout in VRC Task 3 (it looks perfectly aligned when we examine in Gazebo), but have never successfully made a connection. We are still trying. I have someone working full-time on just trying different mating configurations. We'll post if we are successful.

**Comment by osrf-migration** [comment]
**Original comment by Chris Atkeson (Bitbucket: [catkeson](https://bitbucket.org/%7B3bdaa162-b747-4fa9-9d6e-a2d5c92933c6%7D/)).**

----------------------------------------

* Edited issue description

**Comment by osrf-migration** [comment]
**Original comment by Chris Atkeson (Bitbucket: [catkeson](https://bitbucket.org/%7B3bdaa162-b747-4fa9-9d6e-a2d5c92933c6%7D/)).**

----------------------------------------
If you get rid of the angular component of coupling_relative_pose, the correct answer is

```
#!c++
 double rotErr_cga
    = (this->drcFireHose.couplingLink->GetWorldPose().rot.GetXAxis() +
       this->drcFireHose.spoutLink->GetWorldPose().rot.GetZAxis()).GetLength();


```
So it is the X axis of the coupling that is the long axis. Its sign needs to be flipped, hence the plus (+)

**Comment by osrf-migration** [comment]
**Original comment by Steve Peters (Bitbucket: [Steven Peters](https://bitbucket.org/%7B2ccfed09-18b8-4921-8d58-15ef01092802%7D/), GitHub: [scpeters](https://github.com/scpeters)).**

----------------------------------------
We made some changes in [pull request #299](https://osrf-migration.github.io/drcsim-gh-pages/#!/osrf/drcsim/pull-requests/299) that may fix this.

**Comment by osrf-migration** [comment]
**Original comment by John Hsu (Bitbucket: [hsu](https://bitbucket.org/%7B0a186eae-abf0-4514-a951-23db5eccc286%7D/), GitHub: [hsu](https://github.com/hsu)).**

----------------------------------------

* changed state from "new" to "resolved"

fixed in [pull request #299](https://osrf-migration.github.io/drcsim-gh-pages/#!/osrf/drcsim/pull-requests/299)
